pennylane/ops/qutrit/non_parametric_ops.py

```diff
@@ -0,0 +1,200 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the ternary discrete-variable quantum operations
+that do not depend on any parameters.
+"""
+# pylint:disable=arguments-differ
+import numpy as np
+
+import pennylane as qml  # pylint: disable=unused-import
+from pennylane.operation import Operation
+
+OMEGA = np.exp(2 * np.pi * 1j / 3)
+ZETA = OMEGA ** (1 / 3)  # ZETA will be used as a phase for later non-parametric operations
+
+# Note: When Operation.matrix() is used for qutrit operations, `wire_order` must be `None` as specifying
```

Comment:
It may be worth overriding the `matrix` function for these operations then to raise an error which describes the issue, if a user tries to pass something in.

Reply:
Thats a good idea. Although I'm a bit hesitant about it since the `matrix` function is meant to be quite general. It might be a better idea to update the function `expand_matrix()` in `pennylane/operation.py` to accept more general dimensions so that wire ordering works for qutrits. For now, that can be done by checking if the base matrix being expanded has an odd number of elements, but once qudit generalization begins, it might be worth adding a `dim` attribute to operations so that dimension-dependent functions can be generalized easily.

Reply:
It's out of scope for this PR I think (so I'd be inclined to just leave the error message). Maybe @antalszava can weigh in here. Personally it's the kind of thing that would make for a nice "good first issue" for a new contributor :smile:

Reply:
How is this case handled now programmatically? I.e., what happens if we do pass in a sensible value for `wire_order`?

It would definitely be great to work it out for the qutrit case at one point since users may want to use wire orders when querying the matrix, especially as iirc this is required when the matrix of the operation is used along with a device that has custom wire labels.

> It's out of scope for this PR I think (so I'd be inclined to just leave the error message). Maybe @antalszava can weigh in here. Personally it's the kind of thing that would make for a nice "good first issue" for a new contributor 

I think either work. :+1: We just have to either allow the feature to work or inform users via an informative error message that this is not something that's supported right now.

Reply:
UPDATE: I just did a local test of `pennylane/operation.py:expand_matrix` using the solution I suggested for checking the dimension of the matrix, and it seems to work well, and only 6 lines of code got changed. I've attached the suggested change below:

```python
def expand_matrix(base_matrix, wires, wire_order=None):
    """Re-express a base matrix acting on a subspace defined by a set of wire labels
    according to a global wire order.

    Args:
        base_matrix (tensor_like): base matrix to expand
        wires (Iterable): wires determining the subspace that base matrix acts on; a base matrix of
            dimension :math:`2^n` or `3^n` acts on a subspace of :math:`n` wires
        wire_order (Iterable): global wire order, which has to contain all wire labels in ``wires``, but can also
            contain additional labels

    Returns:
        tensor_like: expanded matrix

    **Example**

    If the wire order is identical to ``wires``, the original matrix gets returned:

    >>> base_matrix = np.array([[1, 2, 3, 4],
    ...                         [5, 6, 7, 8],
    ...                         [9, 10, 11, 12],
    ...                         [13, 14, 15, 16]])
    >>> print(expand_matrix(base_matrix, wires=[0, 2], wire_order=[0, 2]))
    [[ 1  2  3  4]
     [ 5  6  7  8]
     [ 9 10 11 12]
     [13 14 15 16]]

    If the wire order is a permutation of ``wires``, the entries of the base matrix get permuted:

    >>> print(expand_matrix(base_matrix, wires=[0, 2], wire_order=[2, 0]))
    [[ 1  3  2  4]
     [ 9 11 10 12]
     [ 5  7  6  8]
     [13 15 14 16]]

    If the wire order contains wire labels not found in ``wires``, the matrix gets expanded:

    >>> print(expand_matrix(base_matrix, wires=[0, 2], wire_order=[0, 1, 2]))
    [[ 1  2  0  0  3  4  0  0]
     [ 5  6  0  0  7  8  0  0]
     [ 0  0  1  2  0  0  3  4]
     [ 0  0  5  6  0  0  7  8]
     [ 9 10  0  0 11 12  0  0]
     [13 14  0  0 15 16  0  0]
     [ 0  0  9 10  0  0 11 12]
     [ 0  0 13 14  0  0 15 16]]

    The method works with tensors from all autodifferentiation frameworks, for example:

    >>> base_matrix_torch = torch.tensor([[1., 2.],
    ...                                   [3., 4.]], requires_grad=True)
    >>> res = expand_matrix(base_matrix_torch, wires=["b"], wire_order=["a", "b"])
    >>> type(res)
    torch.Tensor
    >>> res.requires_grad
    True

    >>> print(expand_matrix(base_matrix, wires=[0, 2]))
    [[ 1  2  3  4]
     [ 5  6  7  8]
     [ 9 10 11 12]
     [13 14 15 16]]

    """

    if (wire_order is None) or (wire_order == wires):
        return base_matrix

    wire_order = Wires(wire_order)
    n = len(wires)
    shape = qml.math.shape(base_matrix)
    batch_dim = shape[0] if len(shape) == 3 else None
    is_odd = (shape[1] if batch_dim else shape[0]) % 2
    dim = 3 if is_odd else 2
    interface = qml.math.get_interface(base_matrix)  # pylint: disable=protected-access

    # operator's wire positions relative to wire ordering
    op_wire_pos = wire_order.indices(wires)

    identity = qml.math.reshape(
        qml.math.eye(dim ** len(wire_order), like=interface), [dim] * (len(wire_order) * 2)
    )
    # The first axis entries are range(n, 2n) for batch_dim=None and range(n+1, 2n+1) else
    axes = (list(range(-n, 0)), op_wire_pos)

    # reshape op.matrix()
    op_matrix_interface = qml.math.convert_like(base_matrix, identity)
    shape = [batch_dim] + [dim] * (n * 2) if batch_dim else [dim] * (n * 2)
    mat_op_reshaped = qml.math.reshape(op_matrix_interface, shape)
    mat_tensordot = qml.math.tensordot(
        mat_op_reshaped, qml.math.cast_like(identity, mat_op_reshaped), axes
    )

    unused_idxs = [idx for idx in range(len(wire_order)) if idx not in op_wire_pos]
    # permute matrix axes to match wire ordering
    perm = op_wire_pos + unused_idxs
    sources = wire_order.indices(wire_order)
    if batch_dim:
        perm = [p + 1 for p in perm]
        sources = [s + 1 for s in sources]

    mat = qml.math.moveaxis(mat_tensordot, sources, perm)
    shape = (
        [batch_dim] + [dim ** len(wire_order)] * 2 if batch_dim else [dim ** len(wire_order)] * 2
    )
    mat = qml.math.reshape(mat, shape)

    return mat
```

Reply:
It looks like a good hotfix, one thing that could be nice is to comment on these lines:
```py
    is_odd = (shape[1] if batch_dim else shape[0]) % 2
    dim = 3 if is_odd else 2
```
I.e., that we're assuming that if `is_odd` is `True`, then we are in the qutrit case. Likely this should also be mentioned in the docstring along with a qutrit example.